### PR TITLE
feat: add AutocompleteTokenfield component

### PR DESCRIPTION
**Pull request by dkoo**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

As suggested in https://github.com/Automattic/newspack-listings/pull/1#discussion_r492030845, adds the AutocompleteTokenfield component from the Homepage Posts block to newspack-components. This component provides an autocomplete search input field for use in the block editor.

After this PR is released and the NPM package updated, we should be able to remove the code from the `newspack-listings` and `newspack-blocks` repos and import the component from NPM instead.

### How to test the changes in this Pull Request:

Not sure how to test the NPM package functionality until this PR has been merged and the package updated on NPM, but you can test that the component works by importing it to an existing editor component in this repo and adding it to the render output, something like this:

```
<AutocompleteTokenField
	key="test"
	tokens={ [] }
	onChange={ value => console.log( value ) }
	fetchSuggestions={ value => console.log( value ) }
	fetchSavedInfo={ value => console.log( value ) }
	label={ 'Test autocomplete search field' }
	help={ __(
		'Begin typing post title, click autocomplete result to select.',
		'newspack-plugin'
	) }
/>
```

The field won't do anything or provide suggestions without actual fetch functions being passed to the props, but seeing the field and being able to interact with it should be enough to confirm that it works as an import from `newspack-components`.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by jeffersonrabb** [comment]
(Adding @thomasguillot as a reviewer since this is a new component)

**Comment by thomasguillot** [comment]
Is it going to be used within the Newspack Plugin? or is it just because it's convenient to then share it with Blocks and Listing?

If we are going to end up using it for Newspack Plugin then it needs styling. If it's just for conveniency then I'm not sure how I feel to add non-needed component to Newspack Plugin.

Also bear in mind that the style in Newspack Plugin is slightly different than in the editor/Gutenberg.

File: assets/components/src/autocomplete-tokenfield/index.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+	/**
+	 * Render.
+	 */
+	render() {
+		const { help, label = '', maxLength } = this.props;
+		const { suggestions, loading } = this.state;
+
+		return (
+			<div className="autocomplete-tokenfield">
```
Comment (adekbadek) [inline assets/components/src/autocomplete-tokenfield/index.js:168]: The classname should be prefixed with `newspack-` as in the other components.

**Review by adekbadek — CHANGES_REQUESTED** [review]

**Comment by dkoo** [comment]
> Is it going to be used within the Newspack Plugin? or is it just because it's convenient to then share it with Blocks and Listing?

My understanding is that because we're using it in both Blocks and Listings, it demonstrates reusability that warrants putting it in the NPM package. I don't know of any plans to use it in the main plugin yet, but it's certainly a future possibility. Other thoughts from @adekbadek?

File: assets/components/src/autocomplete-tokenfield/index.js (hunk 1 above)
Comment (dkoo) [reply assets/components/src/autocomplete-tokenfield/index.js]: Done in b29c0ce.

File: assets/components/src/autocomplete-tokenfield/index.js [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+				<FormTokenField
+					value={ this.getTokens() }
+					suggestions={ suggestions }
+					onChange={ tokens => this.handleOnChange( tokens ) }
+					onInputChange={ input => this.debouncedUpdateSuggestions( input ) }
+					label={ label }
+					maxLength={ maxLength }
+				/>
+				{ loading && <Spinner /> }
+				{ help && <p className="autocomplete-tokenfield__help">{ help }</p> }
```
Comment (adekbadek) [inline assets/components/src/autocomplete-tokenfield/index.js:178]: The classname should also be updated here ;)

File: assets/components/src/autocomplete-tokenfield/style.scss [hunk 3]
```diff
@@ -0,0 +1,32 @@
+.autocomplete-tokenfield {
```
Comment (adekbadek) [inline assets/components/src/autocomplete-tokenfield/style.scss:1]: The classname should also be updated here, and on line 22

**Review by adekbadek — CHANGES_REQUESTED** [review]

**Comment by adekbadek** [comment]
> Is it going to be used within the Newspack Plugin? or is it just because it's convenient to then share it with Blocks and Listing?

We have no other place to put shared JS code, and the package is called `newspack-components`. Not using `AutocompleteTokenField` in Newspack Plugin might be an argument for separating `newspack-components` to a repository, but I'm not sure it's worth the effort.

**Review by adekbadek — APPROVED** [review]

File: assets/components/src/autocomplete-tokenfield/index.js (hunk 2 above)
Comment (dkoo) [reply assets/components/src/autocomplete-tokenfield/index.js]: Doh!

File: assets/components/src/autocomplete-tokenfield/style.scss (hunk 3 above)
Comment (dkoo) [reply assets/components/src/autocomplete-tokenfield/style.scss]: Doh!

**Comment by dkoo** [comment]
@thomasguillot Do you have any other thoughts or objections about this change? We already have three instances in other plugins where this could be used:

* Newsletters
* Blocks
* Listings

This is a really useful component that I see getting more use in other Newspack plugins. I also can see the argument for making `newspack-components` its own repository, but I'll defer to you and @adekbadek on that.

**Review by thomasguillot — APPROVED** [review]
@dkoo let's add it to the plugin for now and we can revisit `newspack-components` later on if we decide to.

**Pull request merged**